tardis/io/logger/logger.py

```diff
@@ -145,22 +95,190 @@ def logging_state(log_level, tardis_config, specific_log_level):
+    """
+    ansi_escape = re.compile(r"\x1B[@-_][0-?]*[ -/]*[@-~]")
+    return ansi_escape.sub("", text)
+
+
+class WidgetHandler(logging.Handler):
+    """
+    A custom logging handler that outputs log messages to IPython widgets.
+
+    Parameters
+    ----------
+    logging.Handler : class
+        Inherits from the logging.Handler class.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def emit(self, record):
+        """
+        Emit a log record.
+
+        Parameters
+        ----------
+        record : logging.LogRecord
+            The log record to be emitted.
+        """
+        log_entry = self.format(record)
+        clean_log_entry = remove_ansi_escape_sequences(log_entry)
+
+        color = LOG_LEVEL_COLORS.get(
+            record.levelno, LOG_LEVEL_COLORS["default"]
+        )
+
+        parts = clean_log_entry.split(" ", 2)
+        if len(parts) > 2:
+            prefix = parts[0]
+            levelname = parts[1]
+            message = parts[2]
+            html_output = f'<span>{prefix}</span> <span style="color: {color}; font-weight: bold;">{levelname}</span> {message}'
+        else:
+            html_output = clean_log_entry
+
+        if record.levelno in (logging.WARNING, logging.ERROR):
+            with log_outputs["WARNING/ERROR"]:
+                display(
+                    HTML(
+                        f"<pre style='white-space: pre-wrap; word-wrap: break-word;'>{html_output}</pre>"
+                    )
+                )
+        elif record.levelno == logging.INFO:
+            with log_outputs["INFO"]:
+                display(
+                    HTML(
+                        f"<pre style='white-space: pre-wrap; word-wrap: break-word;'>{html_output}</pre>"
+                    )
+                )
+        elif record.levelno == logging.DEBUG:
+            with log_outputs["DEBUG"]:
+                display(
+                    HTML(
+                        f"<pre style='white-space: pre-wrap; word-wrap: break-word;'>{html_output}</pre>"
+                    )
+                )
+        with log_outputs["ALL"]:
+            display(
+                HTML(
+                    f"<pre style='white-space: pre-wrap; word-wrap: break-word;'>{html_output}</pre>"
+                )
+            )
+
+
+widget_handler = WidgetHandler()
+widget_handler.setFormatter(
+    logging.Formatter(
+        "%(name)s [%(levelname)s] %(message)s (%(filename)s:%(lineno)d)"
+    )
+)
+
+logging.captureWarnings(True)
+logger = logging.getLogger("tardis")
+logger.setLevel(logging.DEBUG)
+
+# To fix the issue of duplicate logs
+for handler in logger.handlers[:]:
+    logger.removeHandler(handler)
+
+root_logger = logging.getLogger()
+for handler in root_logger.handlers[:]:
+    root_logger.removeHandler(handler)
+
+logger.addHandler(widget_handler)
+logging.getLogger("py.warnings").addHandler(widget_handler)
+create_and_display_log_tab(log_outputs)
```

Comment:
The lines 233-254 are strange because they are not in a class or function. Why are they set up like this? There seems to be a lot of calls to logger and logging.